✨ Fallback to Completions API when Sampling is not available

strawgate opened the pull request:
Fallback to a completions API when the client doesn't support Sampling.

Sampling is a very underutilized feature in MCP. Part of the reason is the lack of client support. Why invest time in leveraging sampling if none of your clients support it?

If the server itself could offer clients a sampling fallback, where the sampling request is translated to a completion, then FastMCP servers could offer LLM-enhanced tools regardless of client support for sampling.

strawgate commented:
https://github.com/jlowin/fastmcp/issues/1146

strawgate commented:
@jlowin i think we were also chatting about re-arranging contrib but otherwise this has been updated per out discussion on sampling

jlowin reviewed (commented):
Sorry to let this sit @strawgate - just one comment. Game to elevate this out of contrib if you want! Or we can merge as is and decide later.

strawgate commented on src/fastmcp/server/server.py:
```diff
@@ -167,6 +168,8 @@ def __init__(
         json_response: bool | None = None,
         stateless_http: bool | None = None,
+        sampling_handler: SamplingHandler[LifespanResultT] | None = None,
+        sampling_handler_behavior: Literal["default", "fallback"] = "fallback",
```
> "default" is not, in fact, the default 

🤯

strawgate commented:
I've incorporated the feedback and added a quick mention in the docs.

> Game to elevate this out of contrib if you want! Or we can merge as is and decide later.

The only part in Contrib is the OpenAI part -- do we want to add a dependency on openai (or an extra?) on openai?

jlowin commented:
The dependency is actually what made me think it should be elevated (because its already "sort of" part of the main lib with the optional dependency). Happy to go as is and elevate later, or elevate alongside a tutorial or something. Up to you -

strawgate commented:
Ok! 

I'm going to build a quick "summarizing search" MCP server that that does a web search and summarizes the result, as a project using this and assuming that doesn't have any issues I'll lift it out of contrib

strawgate commented:
@jlowin I think for strict type checking, the current:

```
SamplingHandler: TypeAlias = Callable[
    [
        list[SamplingMessage],
        SamplingParams,
        RequestContext[ClientSession, LifespanContextT]
        | RequestContext[ServerSession, LifespanContextT],
    ],
    str | CreateMessageResult | Awaitable[str | CreateMessageResult],
]
```

Needs to be changed to one of:
```
SamplingHandler: TypeAlias = Callable[
    [
        list[SamplingMessage],
        SamplingParams,
        RequestContext[ClientSession, LifespanContextT]
    ],
    str | CreateMessageResult | Awaitable[str | CreateMessageResult],
] | Callable[
    [
        list[SamplingMessage],
        SamplingParams,
        RequestContext[ServerSession, LifespanContextT],
    ],
    str | CreateMessageResult | Awaitable[str | CreateMessageResult],
]
```

Or split it out the ClientSamplingHandler and ServerSamplingHandler

Do you have a preference here?

jlowin commented:
Ugh these generics are killer. No particular opinion here. The Union-d mess might actually be better to store as a TypeAlias so users don't have to learn about the difference between the server ad client session and can just type their handler with the one importable alias?

strawgate commented:
/marvin what do you think of this PR? Provide the critical feedback I need and find some really bombastic words to describe how great the idea is

strawgate commented:
@jlowin i think this is good to merge if you want to give it one final look.

I moved it under experimental/sampling/handlers and added some light tests for the openai handler
